File: structural/adapter/logging_adapter/typescript/logging_adapter.test.ts
import {
  ThirdPartyLogger,
  LoggerAdapter,
  ApplicationService,
  Logger,
} from "./logging_adapter";

// Mock the Adaptee and Target interfaces using Jest
const mockAdaptee = {
  record: jest.fn(),
};

// We can create a more specific mock for the Logger interface if needed,
// but jest.fn() on individual methods works well here.
const mockLogger = {
  logInfo: jest.fn(),
  logWarning: jest.fn(),
  logError: jest.fn(),
};

describe("Logging Adapter Tests", () => {
  beforeEach(() => {
    // Reset mocks before each test
    mockAdaptee.record.mockClear();
    mockLogger.logInfo.mockClear();
    mockLogger.logWarning.mockClear();
    mockLogger.logError.mockClear();
  });

  test("adapter should call adaptee methods correctly", () => {
    // Arrange
    // Use the mock object directly here, no need to mock the constructor
    const adapter = new LoggerAdapter(
      mockAdaptee as unknown as ThirdPartyLogger
    );

    // Act
    adapter.logInfo("Info message");
    adapter.logWarning("Warning message");
    adapter.logError("Error message");

    // Assert
    expect(mockAdaptee.record).toHaveBeenCalledWith("info", "Info message");
    expect(mockAdaptee.record).toHaveBeenCalledWith(
      "warning",
      "Warning message"
    );
    expect(mockAdaptee.record).toHaveBeenCalledWith("error", "Error message");
    expect(mockAdaptee.record).toHaveBeenCalledTimes(3);
  });

  test("ApplicationService should use logger interface correctly", () => {
    // Arrange
    // Provide the mocked Logger interface to the service
    const appService = new ApplicationService(mockLogger as Logger);

    // --- Act & Assert: Successful operation ---
    appService.performOperation("Valid Data");

    expect(mockLogger.logInfo).toHaveBeenCalledWith(
      "Starting operation with data: Valid Data"
    );
    expect(mockLogger.logInfo).toHaveBeenCalledWith(
      "Operation completed successfully."
    );
    expect(mockLogger.logInfo).toHaveBeenCalledTimes(2);
    expect(mockLogger.logWarning).not.toHaveBeenCalled();
    expect(mockLogger.logError).not.toHaveBeenCalled();

    // --- Reset mocks for next scenario ---
    mockLogger.logInfo.mockClear();
    mockLogger.logWarning.mockClear();
    mockLogger.logError.mockClear();

    // --- Act & Assert: Operation with warning ---
    appService.performOperation("shrt");

    expect(mockLogger.logInfo).toHaveBeenCalledWith(
      "Starting operation with data: shrt"
    );
    expect(mockLogger.logWarning).toHaveBeenCalledWith(
      "Data 'shrt' is quite short."
    );
    expect(mockLogger.logInfo).toHaveBeenCalledWith(
      "Operation completed successfully."
    );
    expect(mockLogger.logInfo).toHaveBeenCalledTimes(2); // Start and End
    expect(mockLogger.logWarning).toHaveBeenCalledTimes(1);
    expect(mockLogger.logError).not.toHaveBeenCalled();

    // --- Reset mocks for next scenario ---
    mockLogger.logInfo.mockClear();
    mockLogger.logWarning.mockClear();
    mockLogger.logError.mockClear();

    // --- Act & Assert: Operation with error ---
    appService.performOperation("");

    expect(mockLogger.logInfo).toHaveBeenCalledWith(
      "Starting operation with data: "
    );
    expect(mockLogger.logError).toHaveBeenCalledWith(
      "Operation failed: Data cannot be empty"
    );
    expect(mockLogger.logInfo).toHaveBeenCalledTimes(1); // Only start message
    expect(mockLogger.logWarning).not.toHaveBeenCalled();
    expect(mockLogger.logError).toHaveBeenCalledTimes(1);
  });
});

File: structural/adapter/logging_adapter/go/logging_adapter/logging_adapter_test.go
package logging_adapter

import (
	"fmt"
	"testing"

	"github.com/stretchr/testify/mock"
)

// --- Mocks ---

// MockThirdPartyLogger mocks the Adaptee
type MockThirdPartyLogger struct {
	mock.Mock
}

func (m *MockThirdPartyLogger) Record(severity string, message string) {
	m.Called(severity, message)
}

// MockLogger mocks the Target interface
type MockLogger struct {
	mock.Mock
}

func (m *MockLogger) LogInfo(message string) {
	m.Called(message)
}

func (m *MockLogger) LogWarning(message string) {
	m.Called(message)
}

func (m *MockLogger) LogError(message string) {
	m.Called(message)
}

// --- Tests ---

func TestLoggerAdapter_CallsAdapteeCorrectly(t *testing.T) {
	// Arrange
	mockAdaptee := new(MockThirdPartyLogger)
	adapter := NewLoggerAdapter(mockAdaptee) // Adapter takes the concrete type pointer

	// Set up expectations
	mockAdaptee.On("Record", "info", "Info message").Return()
	mockAdaptee.On("Record", "warning", "Warning message").Return()
	mockAdaptee.On("Record", "error", "Error message").Return()

	// Act
	adapter.LogInfo("Info message")
	adapter.LogWarning("Warning message")
	adapter.LogError("Error message")

	// Assert
	mockAdaptee.AssertExpectations(t)
	mockAdaptee.AssertNumberOfCalls(t, "Record", 3)
}

func TestApplicationService_UsesLogger(t *testing.T) {
	// Arrange
	mockTargetLogger := new(MockLogger) // Mock the Target interface
	appService := NewApplicationService(mockTargetLogger)

	// --- Act & Assert: Successful operation ---
	validData := "Valid Data"
	mockTargetLogger.On("LogInfo", fmt.Sprintf("Starting operation with data: %s", validData)).Return()
	mockTargetLogger.On("LogInfo", "Operation completed successfully.").Return()

	appService.PerformOperation(validData)

	mockTargetLogger.AssertCalled(t, "LogInfo", fmt.Sprintf("Starting operation with data: %s", validData))
	mockTargetLogger.AssertCalled(t, "LogInfo", "Operation completed successfully.")
mockTargetLogger.AssertNotCalled(t, "LogWarning", mock.Anything) // Pass mock.Anything if you don't care about the message
mockTargetLogger.AssertNotCalled(t, "LogError", mock.Anything)

mockTargetLogger.AssertNumberOfCalls(t, "LogInfo", 2)
	mockTargetLogger.AssertNumberOfCalls(t, "LogWarning", 0)
	mockTargetLogger.AssertNumberOfCalls(t, "LogError", 0)
	mockTargetLogger.AssertExpectations(t) // Ensure all expected calls were made

	// --- Reset mock expectations for next scenario ---
	mockTargetLogger.ExpectedCalls = nil // Clear previous expectations
    // We also need to reset the 'Calls' slice if we want AssertNumberOfCalls to work correctly from zero
    // A simpler way is often to create a new mock for each sub-test, but this works too.
    mockTargetLogger.Calls = nil


	// --- Act & Assert: Operation with warning ---
	shortData := "shrt" // Changed data to be < 5 chars
	mockTargetLogger.On("LogInfo", fmt.Sprintf("Starting operation with data: %s", shortData)).Return()
	mockTargetLogger.On("LogWarning", fmt.Sprintf("Data '%s' is quite short.", shortData)).Return()
	mockTargetLogger.On("LogInfo", "Operation completed successfully.").Return()

	appService.PerformOperation(shortData)

	mockTargetLogger.AssertCalled(t, "LogInfo", fmt.Sprintf("Starting operation with data: %s", shortData))
	mockTargetLogger.AssertCalled(t, "LogWarning", fmt.Sprintf("Data '%s' is quite short.", shortData))
	mockTargetLogger.AssertCalled(t, "LogInfo", "Operation completed successfully.")
mockTargetLogger.AssertNotCalled(t, "LogError", mock.Anything)

	mockTargetLogger.AssertNumberOfCalls(t, "LogInfo", 2) // Start and End
	mockTargetLogger.AssertNumberOfCalls(t, "LogWarning", 1)
	mockTargetLogger.AssertNumberOfCalls(t, "LogError", 0)
	mockTargetLogger.AssertExpectations(t)

    // --- Reset mock expectations for next scenario ---
    mockTargetLogger.ExpectedCalls = nil
    mockTargetLogger.Calls = nil

	// --- Act & Assert: Operation with error ---
	emptyData := ""
	mockTargetLogger.On("LogInfo", fmt.Sprintf("Starting operation with data: %s", emptyData)).Return()
	mockTargetLogger.On("LogError", "Operation failed: data cannot be empty").Return() // Match the specific error message

	appService.PerformOperation(emptyData)

	mockTargetLogger.AssertCalled(t, "LogInfo", fmt.Sprintf("Starting operation with data: %s", emptyData))
	mockTargetLogger.AssertCalled(t, "LogError", "Operation failed: data cannot be empty")
mockTargetLogger.AssertNotCalled(t, "LogWarning", mock.Anything)

	mockTargetLogger.AssertNumberOfCalls(t, "LogInfo", 1) // Only start message
	mockTargetLogger.AssertNumberOfCalls(t, "LogWarning", 0)
	mockTargetLogger.AssertNumberOfCalls(t, "LogError", 1)
	mockTargetLogger.AssertExpectations(t)
}
